### PR TITLE
Implement selective subtree pruning

### LINKED ISSUE #4117: Slow hierarchy merging when there's many unchanged folders
Some recent scale tests show inefficiencies in hierarchy merging for incremental backups if there are many unchanged folders in the backup. We can increase the efficiency of this by skipping merging on unchanged or completely deleted subtrees of the hierarchy

The snapshot operation will continue to traverse those parts of the hierarchy, but that can be optimized separately

### PR DESCRIPTION
**Pull request by ashmrtn**
Determine if a subtree has any directory object changes and if it
doesn't skip traversing that subtree during hierarchy merging.
"Directory object changes" for a subtree are defined as:
  * moving/renaming a directory within the subtree
  * deleting a directory in the subtree
  * moving a directory into the subtree
  * creating a directory in the subtree
  * moving a directory out of the subtree

The resulting snapshot still contains all data in the
pruned subtree.

---

#### Does this PR need a docs update or release note?

- [x] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [ ] :no_entry: No

#### Type of change

- [x] :sunflower: Feature
- [ ] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Supportability/Tests
- [ ] :computer: CI/Deployment
- [ ] :broom: Tech Debt/Cleanup

#### Issue(s)

* #4117

#### Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Comment by pandeyabs** [comment]
Sorry, haven't had a chance to review the proposal and PR yet. Will do it in the next 1-2 days.

File: src/internal/kopia/upload.go [hunk 1]
```diff
@@ -1061,6 +1156,9 @@ const (
 	// statSkipMerge denotes the number of directories that weren't merged because
 	// they were marked either DoNotMerge or New.
 	statSkipMerge = "directories_skipped_merging"
+	// statPruned denotes the number of subtree roots that selective subtree
+	// pruning applied to.
+	statPruned = "subtrees_pruned"
```
Comment (pandeyabs) [inline src/internal/kopia/upload.go:1161]: nit: Should we add a TODO here? I don't see `statPruned` being modified in this PR.

File: src/internal/kopia/upload_test.go [hunk 2]
```diff
@@ -3017,3 +3025,783 @@ func (suite *HierarchyBuilderUnitSuite) TestBuildDirectoryTreeSelectsMigrateSubt
...
+	table := []struct {
+		name                  string
+		inputCollections      func(t *testing.T) []data.BackupCollection
+		inputExcludes         *pmMock.PrefixMap
+		expected              *expectedNode
+		expectedIterateCounts map[string]int
+	}{
+		{
+			// Test that even if files are excluded in the subtree selective subtree
+			// pruning skips traversing the it during hierarchy merging and the file
```
Comment (pandeyabs) [inline src/internal/kopia/upload_test.go:3260]: nit: Could you improve this comment a bit?

**Review by pandeyabs — APPROVED** [review]
Looks good! Test cases are solid.

File: src/internal/kopia/upload.go (hunk 1 above)
Comment (ashmrtn) [reply src/internal/kopia/upload.go:1163]: oh whoops, good catch! I'll just add that in real quick. It'll just be another branch off one of the if-blocks

File: src/internal/kopia/upload_test.go (hunk 2 above)
Comment (ashmrtn) [reply src/internal/kopia/upload_test.go]: sure thing!

**Pull request merged**